library/init/meta/match_tactic.lean

```diff
@@ -7,37 +7,65 @@ prelude
 import init.meta.interactive_base init.function
 
 namespace tactic
+/-- A pattern is an expression `target` containing temporary metavariables.
+A pattern also contains a list of `outputs` which also depend on these temporary metavariables.
+When we run `match p e`, the system will match `p.target` with `e` and assign
+the temporary metavariables. It then returns the outputs with the assigned variables.
+
+## Fields
+
+- `target` Term to match. Contains temporary metavariables.
+- `uoutput` List of universes that are returned on a successful match.
+- `moutput` List of expressions that are returned on a successful match.
+- `nuvars` Number of (temporary) universe metavariables in this pattern.
+- `nmvars` Number of (temporary) metavariables in this pattern.
+
+## Example
+
+The pattern for `list.cons h t` returning `h` and `t` would be
+```
+{ target  := `(@list.cons ?x_0 ?x_1 ?x_2),
+  uoutput := [],
+  moutput := [?x_1,?x_2],
+  nuvars  := 0,
+  nmvars  := 3
+}
+```
+
+-/
 meta structure pattern :=
-/- Term to match. -/
 (target : expr)
-/- Set of universes that is instantiated for each successful match. -/
 (uoutput : list level)
-/- Set of terms that is instantiated for each successful match. -/
 (moutput : list expr)
-/- Number of (temporary) universe meta-variables in this pattern. -/
 (nuvars : nat)
-/- Number of (temporary) meta-variables in this pattern. -/
 (nmvars : nat)
 
-/-- `mk_pattern ls es t u o` creates a new pattern with (length ls) universe meta-variables and (length es) meta-variables.
-   In the produced pattern p, we have that
-   - `pattern.target p` is the term t where the universes ls and expressions es have been replaced with temporary meta-variables.
-   - `pattern.uoutput p` is the list u where the universes ls have been replaced with temporary meta-variables.
-   - `pattern.moutput p` is the list o where the universes ls and expressions es have been replaced with temporary meta-variables.
-   - `pattern.nuvars p` = length ls
-   - `pattern.nmvars p` = length es
+/-- `mk_pattern umetas emetas target uoutput eoutput` creates a new pattern. The arguments are
+- `umetas` a list of level params to be replaced with temporary universe metavariables.
+- `emetas` a list of local constants to be replaced with temporary metavariables.
+- `target` the expression to be matched.
+- `uoutput` a list of levels to return upon matching. Fails if this depends on metavariables.
+- `eoutput` a list of expressions to return upon matching. Fails if this depends on metavariables.
 
-   The tactic fails if o and the types of es do not contain all universes ls and expressions es. -/
-meta constant mk_pattern : list level → list expr → expr → list level → list expr → tactic pattern
+The procedure is as follows:
+1. Convert all metavariables in `target` to temporary metavariables.
```

Comment:
because temporary meta variables are not an accessible concept in the tactic language, I'm hesitant to mention them in the documentation.

Reply:
Could it be that this should be a design note?

Reply:
Temporary metas aren't accessible in the tactic language but they are accessible on `pattern`. The fields of `pattern` make no sense if you can't talk about temporary metavariables. `match_tactic` is one of the few places in lean that lets you create temporary metavaribles.

Reply:
Does it mixes temporary and non-temporary meta vars? If not, it might be simpler to document this interface by saying that `match_tactic` treats meta vars differently from everywhere else.

Reply:
It does not mix regular and temp metas. I  think it is more confusing if we pretend that tmp metas don't exist.

Reply:
I am happy to add some extra docs explaining what temporary metas are.